dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,199 @@
+ * Downsides:
+ * <ul>
+ *     <li>a single high volume client can result in unbalanced utilization of server nodes.
+ *     <li>when a node is restarted (and clients see 500s/connect exceptions), they will _all_ pin to a different node
+ * </ul>
+ *
+ * To alleviate the second downside, we reshuffle all nodes every 10 minutes.
+ */
+public final class PinUntilErrorChannel implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(PinUntilErrorChannel.class);
+
+    // we also add some jitter to ensure that there isn't a big spike of reshuffling every 10 minutes.
+    private static final Duration RESHUFFLE_EVERY = Duration.ofMinutes(10);
+
+    // TODO(dfox): could we make this endpoint-specific?
+    private final AtomicInteger currentHost = new AtomicInteger(0); // always positive
+    private final Optional<Reshuffle> reshuffle;
+
+    private volatile ImmutableList<LimitedChannel> channels;
+
+    @VisibleForTesting
+    PinUntilErrorChannel(List<LimitedChannel> channels, Random random, Ticker clock) {
+        this.channels = shuffleImmutableList(channels, random);
+        this.reshuffle = Optional.of(new Reshuffle(random, clock));
+    }
+
+    @VisibleForTesting
+    PinUntilErrorChannel(List<LimitedChannel> channels, Random random) {
+        this.channels = shuffleImmutableList(channels, random);
+        this.reshuffle = Optional.empty();
+    }
+
+    public static PinUntilErrorChannel pinUntilError(List<LimitedChannel> channels) {
+        return new PinUntilErrorChannel(channels, ThreadLocalRandom.current(), System::nanoTime);
+    }
+
+    /**
+     * This strategy results in unbalanced server utilization, especially after an upstream restarts.
+     * @deprecated prefer {@link #pinUntilError}
+     */
+    @Deprecated
+    public static PinUntilErrorChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+        return new PinUntilErrorChannel(channels, ThreadLocalRandom.current());
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        if (channels.isEmpty()) {
+            log.debug("Rejecting request due to no delegates");
+            return Optional.empty();
+        }
+
+        if (reshuffle.isPresent()) {
+            reshuffle.get().reshuffleChannelsIfNecessary();
+        }
+
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = channels.get(currentIndex % channels.size());
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = incrementCurrentHost(currentIndex);
+            log.debug(
+                    "Current channel rejected request, switching to next channel",
+                    SafeArg.of("currentIndex", currentIndex),
+                    UnsafeArg.of("current", channel),
+                    SafeArg.of("nextIndex", next));
```

Comment:
nit: if log.isDebugEnabled

I prefer guarding trace and debug statements with logsafe args to prevent allocation in the default case.

Reply:
Do we also want to check that next is present?

Reply:
Eh, I think it's fine to get a safearg with an optionalint.empty in it because it emphasises that we got an out of order response.